### PR TITLE
Period souce and delay node

### PR DESCRIPTION
**Pull request by arkadius**

### CODE REVIEW

File: engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/DelayTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import pl.touk.nussknacker.engine.api
+import pl.touk.nussknacker.engine.api._
+import pl.touk.nussknacker.engine.flink.api.compat.ExplicitUidInOperatorsSupport
+import pl.touk.nussknacker.engine.flink.api.process.{FlinkCustomNodeContext, FlinkCustomStreamTransformation}
+import pl.touk.nussknacker.engine.flink.util.keyed.StringKeyOnlyMapper
+
+object DelayTransformer extends CustomStreamTransformer with ExplicitUidInOperatorsSupport {
+
+  @MethodToInvoke(returnType = classOf[Void])
+  def invoke(@ParamName("key") key: LazyParameter[CharSequence],
```
Comment (mproch) [inline engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/DelayTransformer.scala:21]: I started to wonder - do we really need the key to be passed by user explicitly? 
We need it internally - but isn't it enough to generate random number or sth like that? 
We used similar element at one of the clients and they struggled to understand what the key should be for this element....

File: engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/DelayTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+class DelayFunction(nodeId: String, delay: Duration)
+  extends KeyedProcessFunction[String, ValueWithContext[String], ValueWithContext[AnyRef]] {
+
+  type StateType = List[TimestampedValue[api.Context]]
+  type FlinkCtx = KeyedProcessFunction[String, ValueWithContext[String], ValueWithContext[AnyRef]]#Context
+  type FlinkTimerCtx = KeyedProcessFunction[String, ValueWithContext[String], ValueWithContext[AnyRef]]#OnTimerContext
+
+  private val descriptor = new ValueStateDescriptor[StateType]("state", classOf[StateType])
```
Comment (mproch) [inline engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/DelayTransformer.scala:43]: Wouldn't ListState be better?

File: engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/PeriodicSourceFactory.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+// TODO: add testing capabilities
+object PeriodicSourceFactory extends PeriodicSourceFactory(new MapAscendingTimestampExtractor(MapAscendingTimestampExtractor.DefaultTimestampField))
+
+class PeriodicSourceFactory(timestampAssigner: TimestampAssigner[jul.Map[String, AnyRef]]) extends FlinkSourceFactory[jul.Map[String, AnyRef]]  {
+
+  @MethodToInvoke
+  def create(@ParamName("period") period: Duration,
+             // TODO: @DefaultValue(1) instead of nullable
+             @ParamName("count") @Nullable nullableCount: Integer,
+             @ParamName("value") value: LazyParameter[jul.Map[String, AnyRef]]): Source[_] = {
```
Comment (mproch) [inline engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/PeriodicSourceFactory.scala:32]: does it have to be map? Or just AnyRef? I know you have MapAscendingTimestampExtractor below, but for constant time field maybe we could have just another field?

File: engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/PeriodicSourceFactory.scala [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      override def timestampAssignerForTest: Option[TimestampAssigner[jul.Map[String, AnyRef]]] = Some(timestampAssigner)
+
+      override val returnType: typing.TypingResult = value.returnType
+
+    }
+  }
+
+}
+
+// CheckpointedFunction to handle checkpoints of underlying stream - not sure if it is need
```
Comment (mproch) [inline engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/PeriodicSourceFactory.scala:67]: I don't think it is needed if we don't actually use state?

File: engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/DelayTransformer.scala (hunk 1 above)
Comment (arkadius) [reply engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/DelayTransformer.scala]: I've made key optional, with default value - empty string. Also used MapState, so reading state won't hurt us even for one key

File: engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/PeriodicSourceFactory.scala (hunk 3 above)
Comment (arkadius) [reply engine/flink/model-util/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/PeriodicSourceFactory.scala]: Good point, changed to AnyRef. Would be nice to have `@Placeholder({:})` which not works now :/

**Pull request merged**